### ISSUE TITLE
Decouple from Node

### ISSUE THREAD

**Issue by keyboardDrummer** [Microsoft/vscode-languageserver-node#354]
Currently `vscode-languageserver-protocol` depends on Node, since it has a small dependency on `vscode-jsonrpc` which depends on Node. I'm using `vscode-languageserver-protocol` inside a browser, but now I'm forced to include Node types during compilation, otherwise compilation fails.

I know this is a ton of work, but it would very sweet if `vscode-languageserver-protocol` was decoupled from Node and just contained the protocol definition itself, which has nothing to do with Node.

The packages would change to something like
`vscode-jsonrpc` (just the JsonRpc interface)
`vscode-jsonrpc-node` (implementation in node)
`vscode-languageserver-protocol` (just the interface, some Node related references are removed)
`vscode-languageserver` (depends on Node)

**Comment by rcjsuen** [comment]
To preserve backwards compatibility I think the name of the original package has to (or at least, it should) stay `vscode-jsonrpc`. Otherwise, you're causing a lot of churn to downstream package maintainers that are depending on the [`vscode-jsonrpc` npm module](http://npmjs.com/package/vscode-jsonrpc).

**Comment by rcjsuen** [comment]
The `main.ts` file in `vscode-languageserver-protocol` seems to re-export everything(?) from the `vscode-jsonrpc` package.

https://github.com/Microsoft/vscode-languageserver-node/blob/db3a0a2854853ea1a3d72c916eda714fac6e87b9/protocol/src/main.ts#L7-L32

This list includes the most obvious Node references such as `StreamMessageReader` and `IPCMessageReader` so I think you're going to need another npm module that contains stuff from [`protocol.js`](https://github.com/Microsoft/vscode-languageserver-node/blob/db3a0a2854853ea1a3d72c916eda714fac6e87b9/protocol/src/protocol.ts) and its dependent TypeScript files.

For illustration purposes and the lack of a better name:
- `vscode-jsonrpc-nonode` (no Node.js dependencies)
- `vscode-jsonrpc` (the original package preserved re-exporting everything from `vscode-jsonrpc-nonode`
- `vscode-languageserver-protocol-nonode` (no Node.js dependencies and depends on `vscode-jsonrpc-nonode`
- `vscode-languageserver-protocol` (the original package preserved re-exporting everything from `vscode-languageserver-protocol-nonode` and also `vscode-jsonrpc` where applicable per above)

**Comment by rcjsuen** [comment]
@keyboardDrummer Personally, I think it's something worth doing. However, I am not the maintainer of these packages so... :)

**Comment by dbaeumer** [comment]
I am open to this and I think it might be easier to achieve. The protocol.ts file is almost clean and the only 4 types it uses from jsonrpc are RequestType, RequestType0, NotificationType, NotificationType0 which don't have a node dependency either. Since TS supports structural typing we might be able to re-declare them in the protocol package to remove the dependency and then may be simply push 2 different npm packages. 

I would appreciate if someone else could give this a try. I have to say it is not high on my priority list.

**Comment by keyboardDrummer** [comment]
Maybe another option is not to create new packages but to only split up existing modules. Move the types `RequestType, RequestType0, NotificationType, NotificationType0` in `json-rpc` from the `main.ts` module into a separate `interface.ts` module, and then have `protocol.ts` import that.

I think that would cause `tsc` to not require node types when compiling an import of `protocol.ts` right?

**Comment by dbaeumer** [comment]
Yes, but then you can't compile against these types when installing the npm protocol module since the stuff from interface.ts will not be packaged.

**Comment by dbaeumer** [comment]
I will close this as out-of-scope since the item didn't get any traction. I am open to reopen if someone wants to work on it.
